### PR TITLE
Add recording data and read from previous data

### PR DESCRIPTION
**Pull request by khang-874**
This will write all datagram to a file with filename be the current date and .dump extension.
It will write in raw bytes, and read according just like when we receive an packet

### CODE REVIEW

File: packet_spec.py [hunk 1]
```diff
@@ -117,6 +117,17 @@ def parse_packet_header(header_bytes: bytes) -> PacketHeader:
...
+    match header.type:
+        case PacketType.TELEMETRY:
+            match header.sub_type:
+                case TelemetryPacketSubType.TEMPERATURE | TelemetryPacketSubType.PRESSURE | TelemetryPacketSubType.MASS:
+                    return 9
+                case TelemetryPacketSubType.ACT_STATE:
+                    return 6
+                case TelemetryPacketSubType.ARMING_STATE | TelemetryPacketSubType.WARNING:
+                    return 5
+
```
Comment (EliasJRH) [inline packet_spec.py:130]: Nothing to change here. Just wanted to say that I believe this is likely how the revamped ground station will calculate packet length and I'm happy to see how clean it looks

File: widget.py [hunk 2]
```diff
@@ -141,6 +144,9 @@ def join_multicast_group(self, ip_addr, port):
             self.ui.udpConnectButton.setText("Close UDP connection")
             self.ui.udpIpAddressInput.setReadOnly(True)
             self.ui.udpPortInput.setReadOnly(True)
+            file_name = QDateTime.currentDateTime().toString("yyyy-MM-dd_HH-mm")
+            file_name += '.dump'
+            self.file_out = open(file_name, "a+b")
```
Comment (EliasJRH) [inline widget.py:149]: I'm good with the file naming, can we put these files in a recordings directory and add that to the gitignore?

File: widget.py [hunk 3]
```diff
@@ -178,15 +184,36 @@ def udp_connection_button_handler(self):
...
+
+    def open_file_button_handler(self):
+            # options = QFileDialog.Options()
+            file_path, _ = QFileDialog.getOpenFileName(self, "Open Previous File", "", "Dump file(*.dump);;All files (*)")
 
+            # If a file is selected, read its contents
+            if file_path:
+                with open(file_path, 'rb') as file:
+                    data = file.read()
+                    self.display_previous_data(data)
```
Comment (EliasJRH) [inline widget.py:209]: Looks good here, don't forget to remove the lines not being used!

**Review by EliasJRH — CHANGES_REQUESTED** [review]
Overall I think you did a great job implementing the playback feature as I imagine that if we want to see a past test, we'd want all the data at once and not just have them be played back at the same speed.

That being said, it might get annoying during development if everything was always being logged. Can you also add a button to toggle when data is being recorded? Also let's maybe move both the playback and record toggle button to underneath the output log in a horizontal layout. We'll probably be adding more buttons under there anyway.

Nice work so far! Just a couple extra things but that's how it be when developing this kind of stuff ¯\_(ツ)_/¯

**Comment by khang-874** [comment]
When we open the previous data it will display every data once.
Also I just add toggle button to record file. I also move the update act state function to the plot point function. Can you check it out ?

File: widget.py [hunk 4]
```diff
@@ -196,11 +235,11 @@ def udp_receive_socket_data(self):
...
-            #Actuator state handling
-            if(header.sub_type == packet_spec.TelemetryPacketSubType.ACT_STATE):
-                self.updateActState(message)
-            else:
-                self.plot_point(header, message)
+            self.plot_point(header, message)
+
+            #If we want to recording data
+            if self.ui.recordingToggleButton.isChecked():
+                self.file_out.write(datagram)
```
Comment (EliasJRH) [inline widget.py:242]: Can you revert this change to how it was previously? I think it's a better idea these changes it separate PRs

**Review by EliasJRH — CHANGES_REQUESTED** [review]
Just that one thing and I think we're good to go, nice work! 👍

File: widget.py (hunk 4 above)
Comment (khang-874) [reply widget.py]: I just change it back already. You can have a look

**Review by EliasJRH — APPROVED** [review]
Looks good, nice work!
